### ISSUE TITLE
JUnit XML Test Reports

### ISSUE THREAD

**Issue by mds1** [runtimeverification/kontrol#329]
JUnit XML test reports are widely used in CI tools to provide more advanced test reporting, such as [CircleCI's Test Insights](https://circleci.com/docs/insights-tests) feature. Having `kontrol prove` support this standard test result format would enable projects to track individual proof failure rates, average individual proof speed, etc., with existing services that support the standard.

**Comment by anvacaru** [comment]
@mds1 Thank you for opening this issue. It would certainly be a nice addition! We'll start by looking at the JUnit XML schema and see what information is required and what information we already have available.

**Comment by lisandrasilva** [comment]
Hi @mds1, 
Thanks a lot for this suggestion, we are working on implementing it. Currently the generated report look like the following:
```xml
<?xml version='1.0' encoding='utf-8'?>
<testsuites tests="4" failures="1" errors="1" time="194.60815978050232" timestamp="2024-02-16 15:07:00.831986">
	<testsuite name="SimpleTest" tests="2" failures="0" errors="0" time="65.24263954162598" timestamp="2024-02-16 15:07:34.434294">
		<testcase name="test_assert_true()" classname="SimpleTest" time="33.59451103210449" />
		<testcase name="testFail_assert_false()" classname="SimpleTest" time="31.648128509521484" />
	</testsuite>
	<testsuite name="AssertTest" tests="1" failures="1" errors="0" time="58.05429410934448" timestamp="2024-02-16 15:09:04.418438">
		<testcase name="test_revert_branch(uint256,uint256)" classname="AssertTest" time="58.05429410934448">
			<failure message="2 Failure nodes. (1 pending and 1 failing)">
Pending nodes: [9]

Failing nodes:

  Node id: 10
  Failure reason:
    Implication check failed, the following is the remaining implication:
    ( ( { true #Equals 0 &lt;=Int CALLER_ID:Int }
    #And ( { true #Equals 0 &lt;=Int ORIGIN_ID:Int }
    #And ( { true #Equals 0 &lt;=Int NUMBER_CELL:Int }
    #And ( { true #Equals 0 &lt;=Int TIMESTAMP_CELL:Int }
    #And ( { true #Equals 0 &lt;=Int VV0_x_114b9705:Int }
    #And ( { true #Equals 0 &lt;=Int VV1_y_114b9705:Int }
    #And ( { true #Equals VV1_y_114b9705:Int &lt;=Int VV0_x_114b9705:Int }
    #And ( { true #Equals CALLER_ID:Int &lt;Int pow160 }
    #And ( { true #Equals ORIGIN_ID:Int &lt;Int pow160 }
    #And ( { true #Equals NUMBER_CELL:Int &lt;=Int maxSInt256 }
    #And ( { true #Equals TIMESTAMP_CELL:Int &lt;Int pow256 }
    #And ( { true #Equals VV0_x_114b9705:Int &lt;Int pow256 }
    #And { true #Equals VV1_y_114b9705:Int &lt;Int pow256 } ) ) ) ) ) ) ) ) ) ) ) ) #Implies { true #Equals foundry_success ( ... statusCode: EVMC_REVERT , failed: #lookup ( .Map , 46308022326495007027972728677917914892729792999299745830475596687180801507328 ) , revertExpected: false , opcodeExpected: false , recordEventExpected: false , eventExpected: false ) } )
  Path condition:
    { true #Equals VV1_y_114b9705:Int &lt;=Int VV0_x_114b9705:Int }
  Model:
    TIMESTAMP_CELL = 0
    VV1_y_114b9705 = 0
    NUMBER_CELL = 0
    VV0_x_114b9705 = 0
    CALLER_ID = 0
    ORIGIN_ID = 0
</failure>
		</testcase>
	</testsuite>
	<testsuite name="MockCallTest" tests="1" failures="0" errors="1" time="71.31122612953186" timestamp="2024-02-16 15:10:15.888966">
		<testcase name="testClearMockCalls()" classname="MockCallTest" time="71.31122612953186">
			<error type="KeyError">Traceback (most recent call last):
  File "/home/user/kontrol/src/kontrol/prove.py", line 253, in init_and_run_proof
    run_prover(
  File "/home/user/evm-semantics/kevm-pyk/src/kevm_pyk/utils.py", line 123, in run_prover
    prover.advance_proof(max_iterations=max_iterations, fail_fast=fail_fast)
  File "/home/user/pyk/src/pyk/proof/proof.py", line 309, in advance_proof
    self.step_proof()
  File "/home/user/pyk/src/pyk/proof/reachability.py", line 730, in step_proof
    self.proof.failure_info = self.failure_info()
  File "/home/user/pyk/src/pyk/proof/reachability.py", line 733, in failure_info
    return APRFailureInfo.from_proof(self.proof, self.kcfg_explore, counterexample_info=self.counterexample_info)
  File "/home/user/pyk/src/pyk/proof/reachability.py", line 808, in from_proof
    _, reason = kcfg_explore.implication_failure_reason(node_cterm, target_cterm)
  File "/home/user/pyk/src/pyk/kcfg/explore.py", line 240, in implication_failure_reason
    consequent_cell = consequent.cell(cell)
  File "/home/user/pyk/src/pyk/cterm.py", line 165, in cell
    return self.cells[cell]
  File "/home/user/pyk/src/pyk/kast/inner.py", line 689, in __getitem__
    return self._subst[key]
  File "/home/user/pyk/src/pyk/utils.py", line 66, in __getitem__
    return self._dict[key]
KeyError: 'MOCKADDRESS_CELL'
</error>
		</testcase>
	</testsuite>
</testsuites>
```
Please let us know if there's something else you think it would be useful to include in the report. Thank you!

**Comment by mds1** [comment]
This looks great to me! The one thing missing that I see in the [example](https://circleci.com/docs/parallelism-faster-jobs/#junit-xml-reports) from the CircleCI docs is the `file` key containing the file path to a test. This would be useful because sometimes two contracts might have the same test name, so it helps disambiguate

**Comment by lisandrasilva** [comment]
@mds1 thank you for your feedback! Notice that in the example I sent, the tests are organized by contracts; each `<testsuite>` corresponds to a contract. But I agree that having the `file` path to a contract can be useful, so I just added it. 

Another thing we can add, if it is useful, is a property tag with the version of Kontrol that was used to run the tests. See this [example](https://github.com/testmoapp/junitxml?tab=readme-ov-file#complete-junit-xml-example).

**Comment by mds1** [comment]
That seems useful to me, could help with insights into how test performance changes by version

**Comment by lisandrasilva** [comment]
@mds1 Thank you! Here is an example of the generated report:
```xml
<testsuites tests="6" failures="1" errors="1" time="321.40929341316223" timestamp="2024-02-22 10:05:27.352269">
	<testsuite name="AssertTest" tests="3" failures="1" errors="0" time="118.05802774429321" timestamp="2024-02-22 10:05:27.352340">
		<properties>
			<property name="Kontrol version" value="0.1.164" />
		</properties>
		<testcase name="test_assert_true()" classname="AssertTest" time="30.683660984039307" file="test/Simple.t.sol" />
		<testcase name="testFail_assert_false()" classname="AssertTest" time="31.41535449028015" file="test/Simple.t.sol" />
		<testcase name="test_revert_branch(uint256,uint256)" classname="AssertTest" time="55.959012269973755" file="test/Simple.t.sol">
			<failure message="2 Failure nodes. (1 pending and 1 failing)">
Pending nodes: [9]

Failing nodes:

  Node id: 10
  Failure reason:
    Implication check failed, the following is the remaining implication:
    ( ( { true #Equals 0 &lt;=Int CALLER_ID:Int }
    #And ( { true #Equals 0 &lt;=Int ORIGIN_ID:Int }
    #And ( { true #Equals 0 &lt;=Int NUMBER_CELL:Int }
    #And ( { true #Equals 0 &lt;=Int TIMESTAMP_CELL:Int }
    #And ( { true #Equals 0 &lt;=Int VV0_x_114b9705:Int }
    #And ( { true #Equals 0 &lt;=Int VV1_y_114b9705:Int }
    #And ( { true #Equals VV1_y_114b9705:Int &lt;=Int VV0_x_114b9705:Int }
    #And ( { true #Equals CALLER_ID:Int &lt;Int pow160 }
    #And ( { true #Equals ORIGIN_ID:Int &lt;Int pow160 }
    #And ( { true #Equals NUMBER_CELL:Int &lt;=Int maxSInt256 }
    #And ( { true #Equals TIMESTAMP_CELL:Int &lt;Int pow256 }
    #And ( { true #Equals VV0_x_114b9705:Int &lt;Int pow256 }
    #And { true #Equals VV1_y_114b9705:Int &lt;Int pow256 } ) ) ) ) ) ) ) ) ) ) ) ) #Implies { true #Equals foundry_success ( ... statusCode: EVMC_REVERT , failed: #lookup ( .Map , 46308022326495007027972728677917914892729792999299745830475596687180801507328 ) , revertExpected: false , opcodeExpected: false , recordEventExpected: false , eventExpected: false ) } )
  Path condition:
    { true #Equals VV1_y_114b9705:Int &lt;=Int VV0_x_114b9705:Int }
  Model:
    TIMESTAMP_CELL = 0
    ORIGIN_ID = 0
    NUMBER_CELL = 0
    VV1_y_114b9705 = 0
    CALLER_ID = 0
    VV0_x_114b9705 = 0
</failure>
		</testcase>
	</testsuite>
	<testsuite name="MockCallTest" tests="1" failures="0" errors="1" time="65.59461069107056" timestamp="2024-02-22 10:05:27.352775">
		<properties>
			<property name="Kontrol version" value="0.1.164" />
		</properties>
		<testcase name="testClearMockCalls()" classname="MockCallTest" time="65.59461069107056" file="test/MockCall.t.sol">
			<error message="Exception" type="KeyError">
Traceback (most recent call last):
   File "/home/lisandra/evm-semantics/kevm-pyk/src/kevm_pyk/utils.py", line 123, in run_prover
    prover.advance_proof(max_iterations=max_iterations, fail_fast=fail_fast)
   File "/home/lisandra/pyk/src/pyk/proof/proof.py", line 309, in advance_proof
    self.step_proof()
   File "/home/lisandra/pyk/src/pyk/proof/reachability.py", line 734, in step_proof
    self.proof.failure_info = self.failure_info()
   File "/home/lisandra/pyk/src/pyk/proof/reachability.py", line 737, in failure_info
    return APRFailureInfo.from_proof(self.proof, self.kcfg_explore, counterexample_info=self.counterexample_info)
   File "/home/lisandra/pyk/src/pyk/proof/reachability.py", line 814, in from_proof
    _, reason = kcfg_explore.implication_failure_reason(node_cterm, target_cterm)
   File "/home/lisandra/pyk/src/pyk/kcfg/explore.py", line 240, in implication_failure_reason
    consequent_cell = consequent.cell(cell)
   File "/home/lisandra/pyk/src/pyk/cterm.py", line 165, in cell
    return self.cells[cell]
   File "/home/lisandra/pyk/src/pyk/kast/inner.py", line 689, in __getitem__
    return self._subst[key]
   File "/home/lisandra/pyk/src/pyk/utils.py", line 66, in __getitem__
    return self._dict[key]
 KeyError: 'MOCKADDRESS_CELL'
</error>
		</testcase>
	</testsuite>
	<testsuite name="Setup2Test" tests="2" failures="0" errors="0" time="137.75665497779846" timestamp="2024-02-22 10:05:27.384991">
		<properties>
			<property name="Kontrol version" value="0.1.164" />
		</properties>
		<testcase name="test_setup()" classname="Setup2Test" time="79.57158207893372" file="test/SetUp2Test.t.sol" />
		<testcase name="testFail_setup()" classname="Setup2Test" time="58.185072898864746" file="test/SetUp2Test.t.sol" />
	</testsuite>
</testsuites>
```

**Comment by mds1** [comment]
Looks great to me, thank you!

**Comment by lisandrasilva** [comment]
It was closed by this #377.
